<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Blast and Small Alignments • Systematics2020</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Blast and Small Alignments">
<meta property="og:description" content="Systematics2020">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">Systematics2020</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/00_Basics_of_R/index.html">R Reminder</a>
    </li>
    <li>
      <a href="../../articles/01_homology_assignment/index.html">Morphological Homology Assignment</a>
    </li>
    <li>
      <a href="../../articles/02_molecular_homology/index.html">Molecular Homology Assignment</a>
    </li>
    <li>
      <a href="../../articles/03_CTMC/index.html">Nucleotide substitution models</a>
    </li>
    <li>
      <a href="../../articles/03_ParsimonyAndModels/index.html">PAUP, parsimony, and nucletoide substitution models</a>
    </li>
    <li>
      <a href="../../articles/03_tutorial_structure/index.html">Setting Up RevBayes</a>
    </li>
    <li>
      <a href="../../articles/04_Discrete_Morphology/05_RB_MCMC_Discrete_Morph.html">Discrete morphology - Models and Tree Inference</a>
    </li>
    <li>
      <a href="../../articles/04_tutorial_structure/index.html">Tutorial Structure</a>
    </li>
    <li>
      <a href="../../articles/05_pandemic_week/index.html">Pandemic Week</a>
    </li>
    <li>
      <a href="../../articles/06_models/index.html">Nucleotide substitution models</a>
    </li>
    <li>
      <a href="../../articles/07_blast/index.html">Blast and Small Alignments</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Blast and Small Alignments</h1>
            
      
      
      <div class="hidden name"><code>index.Rmd</code></div>

    </div>

    
    
<p>At the start of this semester, we discussed molecular and morphological homology assignment. Today we’ll talk about assigning molecular homology in small-scale datasets, such as single or multi-locus Sanger sequenced data. The nice thing about these data is that they are available for free, and pretty abundantly though GenBank.</p>
<p>That’s great - we have a free and clear data source that you can use with no restrictions. The tricky bit is that installation for using BLAST is a little tricky. We’re going to first install an R Package called <code>PhylotaR</code>, which is mostly used for doing taxonomically-informed searches of GenBank - i.e. most of the stuff we are interested in as population bioloigists.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"remotes"</span>)</pre></body></html></div>
<pre><code>## Installing package into '/tmp/Rtmp2Do8FS/temp_libpath5e856e37248a'
## (as 'lib' is unspecified)</code></pre>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"taxize"</span>)</pre></body></html></div>
<pre><code>## Installing package into '/tmp/Rtmp2Do8FS/temp_libpath5e856e37248a'
## (as 'lib' is unspecified)</code></pre>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="kw pkg">remotes</span><span class="kw ns">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span>(<span class="st">'dombennett/treeman'</span>)</pre></body></html></div>
<pre><code>## Skipping install of 'treeman' from a github remote, the SHA1 (d010b485) has not changed since last install.
##   Use `force = TRUE` to force installation</code></pre>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="kw pkg">remotes</span><span class="kw ns">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span>(<span class="st">'ropensci/outsider'</span>)</pre></body></html></div>
<pre><code>## Skipping install of 'outsider' from a github remote, the SHA1 (f274e30c) has not changed since last install.
##   Use `force = TRUE` to force installation</code></pre>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="kw pkg">remotes</span><span class="kw ns">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span>(<span class="st">'ropensci/phylotaR'</span>, <span class="kw">dependencies</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<pre><code>## Skipping install of 'phylotaR' from a github remote, the SHA1 (faa8b8f5) has not changed since last install.
##   Use `force = TRUE` to force installation</code></pre>
<p>You’ll also need to install the <a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">Blast+ Software</a>. Choose your operating system and the software should walk you through the install.</p>
<p>For the later parts of the lab, please install an aligner (<a href="https://mafft.cbrc.jp/alignment/software/source.html">MAFFT</a>) and a phylogeny maximum likelihood estimator, <a href="">RAXML</a></p>
<p>Once these are installed, we can start looking for sequences. The first thing we need to do is get a GenBank ID key. To do this, we will use the <code>taxize</code> package. The following function pulls up NCBI so you can register.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="kw pkg">taxize</span><span class="kw ns">::</span><span class="fu"><a href="https://docs.ropensci.org/taxize/reference/key_helpers.html">use_entrez</a></span>()</pre></body></html></div>
<pre><code>## Create your key from your (brand-new) account's. 
##  After generating your key set it as ENTREZ_KEY in .Renviron.
##  ENTREZ_KEY='youractualkeynotthisstring'
##  For that, use usethis::edit_r_environ()</code></pre>
<p>Register a user name, then click on your username to be brought to your profile. On the left-hand side, you can generate an API key for unlimited data access. Next, we will add this to our R Environment file so that R will always know we are us when we try to access data.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="kw pkg">usethis</span><span class="kw ns">::</span><span class="fu"><a href="https://usethis.r-lib.org/reference/edit.html">edit_r_environ</a></span>(<span class="kw">scope</span> <span class="kw">=</span> <span class="st">"user"</span>)</pre></body></html></div>
<pre><code>## ● Edit '/home/april/.Renviron'</code></pre>
<pre><code>## ● Restart R for changes to take effect</code></pre>
<p>This will open a file called Renviron. Into it, enter:</p>
<pre><code>ENTREZ_key = &lt;your key number here&gt;</code></pre>
<p>Save, and restart R.</p>
<p>Next, we’ll get the GenBank taxon identifier for a genus of salamanders, <em>Eurycea</em>. Over the nearly 200 years since Linneaus described the specific binomial, there have been a lot of species defined. GenBank uses a universal ID system so that if a genus refers to a plant and an animal genus, you can still find the correct one.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">taxize</span>)
<span class="fu"><a href="https://docs.ropensci.org/taxize/reference/get_uid.html">get_uid</a></span>(<span class="kw">sciname</span> <span class="kw">=</span> <span class="st">"Eurycea"</span>)</pre></body></html></div>
<pre><code>## ══  1 queries  ═══════════════</code></pre>
<pre><code>## 
## Retrieving data for taxon 'Eurycea'</code></pre>
<pre><code>## ✓  Found:  Eurycea
## ══  Results  ═════════════════
## 
## ● Total: 1 
## ● Found: 1 
## ● Not Found: 0</code></pre>
<pre><code>## [1] "43043"
## attr(,"class")
## [1] "uid"
## attr(,"match")
## [1] "found"
## attr(,"multiple_matches")
## [1] FALSE
## attr(,"pattern_match")
## [1] FALSE
## attr(,"uri")
## [1] "https://www.ncbi.nlm.nih.gov/taxonomy/43043"</code></pre>
<p>You’ll see that the unique identifier for <em>Eurycea</em> is 43043. We use this information to set up our GenBank search:</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">phylotaR</span>)
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span>(<span class="st">'eurycea'</span>)</pre></body></html></div>
<pre><code>## Warning in dir.create("eurycea"): 'eurycea' already exists</code></pre>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="fu"><a href="https://docs.ropensci.org/phylotaR/reference/setup.html">setup</a></span>(<span class="kw">wd</span> <span class="kw">=</span> <span class="st">'eurycea'</span>, <span class="kw">txid</span> <span class="kw">=</span> <span class="fl">43043</span>, <span class="kw">ncbi_dr</span> <span class="kw">=</span> <span class="st">'/home/april/software/blast/bin'</span>, <span class="kw">overwrite</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="fu"><a href="https://docs.ropensci.org/phylotaR/reference/run.html">run</a></span>(<span class="kw">wd</span> <span class="kw">=</span> <span class="st">'eurycea'</span>)</pre></body></html></div>
<pre><code>## ... Taxise
## ... Download
## ... Cluster
## ... Cluster2</code></pre>
<p>The above steps will create a directory called <em>Eurycea</em>, set the GenBank parameters and search for sequences. This step will run for a few minutes. The setup step provides the taxon ID and where the BLAST software lives. <code>Run</code> actually completes the search.</p>
<p>On my machine, this took about 6 minutes. Once you see the output <code>Download</code>, you know it’s running. Feel free to take a 10 minute break then.</p>
<p>These data can be quite large, so PhylotaR doesn’t immediately read in the sequences. We use the <code>read_phylota</code> function to get access to the data:</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="no">eury</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/phylotaR/reference/read_phylota.html">read_phylota</a></span>(<span class="st">"eurycea"</span>)</pre></body></html></div>
<p>Let’s take a look and see what species we have in here:</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="no">txids</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/phylotaR/reference/get_txids.html">get_txids</a></span>(<span class="no">eury</span>, <span class="kw">txids</span> <span class="kw">=</span> <span class="no">eury</span>@<span class="kw">txids</span>, <span class="kw">rnk</span> <span class="kw">=</span> <span class="st">'species'</span>)
<span class="co"># First, we get the taxon IDs</span>
<span class="no">txids</span> <span class="kw">&lt;-</span> <span class="no">txids</span>[<span class="no">txids</span> <span class="kw">!=</span>  <span class="st">''</span>]
<span class="co"># And we drop any specimens that are not identified to the species level</span>
<span class="no">txids</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">txids</span>)
<span class="co">#Then we figure out how many uniques we have</span>
<span class="no">txnms</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/phylotaR/reference/get_tx_slot.html">get_tx_slot</a></span>(<span class="no">eury</span>, <span class="no">txids</span>, <span class="kw">slt_nm</span> <span class="kw">=</span> <span class="st">'scnm'</span>)
<span class="co"># And how many of each unique name we have</span>
<span class="no">p</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/phylotaR/reference/plot_phylota_treemap.html">plot_phylota_treemap</a></span>(<span class="kw">phylota</span> <span class="kw">=</span> <span class="no">eury</span>, <span class="kw">txids</span> <span class="kw">=</span> <span class="no">txids</span>, <span class="kw">txnms</span> <span class="kw">=</span> <span class="no">txnms</span>,
                          <span class="kw">area</span> <span class="kw">=</span> <span class="st">'nsq'</span>, <span class="kw">fill</span> <span class="kw">=</span> <span class="st">'ncl'</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">p</span>)</pre></body></html></div>
<p><img src="index_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>We can see some species are better represented than others. The next thing we probably want to do is see what loci we have.</p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="co"># Reduce down to one sequence per taxon</span>
<span class="no">reduced</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/phylotaR/reference/drop_by_rank.html">drop_by_rank</a></span>(<span class="kw">phylota</span> <span class="kw">=</span> <span class="no">eury</span>, <span class="kw">rnk</span> <span class="kw">=</span> <span class="st">'species'</span>, <span class="kw">n</span> <span class="kw">=</span> <span class="fl">1</span>)
<span class="co"># We're going to look at the first cluster first</span>
<span class="no">cid</span> <span class="kw">&lt;-</span> <span class="no">reduced</span>@<span class="kw">cids</span><span class="kw">[[</span><span class="fl">2</span>]]
<span class="co"># extract its sequence IDs from Phylota object</span>
<span class="no">sids</span> <span class="kw">&lt;-</span> <span class="no">reduced</span><span class="kw">[[</span><span class="no">cid</span>]]@<span class="kw">sids</span>
<span class="co"># design sequence names for definition line</span>
<span class="no">sid_txids</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/phylotaR/reference/get_sq_slot.html">get_sq_slot</a></span>(<span class="kw">phylota</span> <span class="kw">=</span>  <span class="no">reduced</span>, <span class="kw">sid</span> <span class="kw">=</span> <span class="no">sids</span>, <span class="kw">slt_nm</span> <span class="kw">=</span> <span class="st">'txid'</span>)
<span class="no">sid_spnms</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/phylotaR/reference/get_tx_slot.html">get_tx_slot</a></span>(<span class="kw">phylota</span> <span class="kw">=</span>  <span class="no">reduced</span>, <span class="kw">txid</span> <span class="kw">=</span> <span class="no">sid_txids</span>, <span class="kw">slt_nm</span> <span class="kw">=</span> <span class="st">'scnm'</span>)
<span class="no">sq_nms</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">sid_spnms</span>, <span class="st">' | '</span>, <span class="no">sids</span>)
<span class="co"># write out</span>
<span class="fu"><a href="https://docs.ropensci.org/phylotaR/reference/write_sqs.html">write_sqs</a></span>(<span class="kw">phylota</span> <span class="kw">=</span>  <span class="no">reduced</span>, <span class="kw">outfile</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="st">"eurycea"</span>, <span class="st">'test.fasta'</span>),
          <span class="kw">sq_nm</span> <span class="kw">=</span> <span class="no">sq_nms</span>, <span class="kw">sid</span> <span class="kw">=</span> <span class="no">sids</span>)</pre></body></html></div>
<p>Now we have a file full of sequences. We probably want to know what these sequences do, though. For this, we can search the annotations. What we want is the definition and feature. This is the EST locus in the large ribsomal subunit. A caution on this is that these can be somewhat non-descript. Sometimes, it is best to search the sequence identifier (like DQ018495.1) on GenBank for more info.</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="no">selected</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/phylotaR/reference/drop_clstrs.html">drop_clstrs</a></span>(<span class="kw">phylota</span> <span class="kw">=</span> <span class="no">reduced</span>, <span class="kw">cid</span> <span class="kw">=</span> <span class="no">cid</span>)
<span class="no">smmry</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">selected</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">smmry</span>)</pre></body></html></div>
<pre><code>##   ID    Type       Seed Parent N_taxa N_seqs Med_sql       MAD
## 1  1 subtree DQ018495.1  43043     13     42    1011 0.5487104
##                Definition Feature
## 1 gene (0.07), and (0.06) est (1)</code></pre>
<p>Crucially, these sequences are unaligned. We will also need to replace the spaces in the taxon names. There are also about ’ characters. Open the <code>test.fasta</code> file. Let’s take a look.</p>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ape</span>)
<span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="st">'mafft --auto eurycea/test.fasta &gt; eurycea/alignment.fasta'</span>)
<span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">'/home/april/software/standard-RAxML/raxmlHPC -m GTRGAMMA -f a -N 10 -p 1234 -x 1234 -n eury -s eurycea/alignment.fasta'</span>))</pre></body></html></div>
<p>Let’s peek at the tree:</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ape</span>)
<span class="no">tree</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span>(<span class="kw">file</span> <span class="kw">=</span> <span class="st">'RAxML_bestTree.eury'</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">tree</span>, <span class="kw">no.margin</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">'unrooted'</span>)</pre></body></html></div>
<p><img src="index_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
